Add Support for count_include_pad in AveragePool in Caffe2 ONNX Backend

hl475 opened the pull request:
Summary:
The goal is to support count_include_pad in Caffe2 ONNX backend. This commit contains the first step - support 4-D tensor cases.
AveragePool with count_include_pad can be expressed as PadImage + AveragePool.

Differential Revision: D8852180

hl475 commented:
@pytorchbot retest this please

hl475 commented:
@houseroad @bddppq, please take a look. Thanks!

houseroad commented on caffe2/onnx/backend.cc:
```diff
@@ -515,6 +515,59 @@ Caffe2Ops Caffe2Backend::CreateConvPoolOpBase(
+      }
+
+      caffe2::Argument arg_pads;
+      arg_pads.add_ints(pads.Get(1));
```
This does not seem right to me. pad[0, 1, 2, 3] should be pad[t, l, b, r]

houseroad commented on caffe2/onnx/backend.cc:
```diff
@@ -515,6 +515,59 @@ Caffe2Ops Caffe2Backend::CreateConvPoolOpBase(
+  if (attributes.HasAttribute("count_include_pad")) {
+    auto count_include_pad = attributes.get<int64_t>("count_include_pad", 0L);
+    if (count_include_pad == 1) {
+      std::string pad_name = opset_version < 2 ? "paddings" : "pads";
```
There are some helper functions in https://github.com/pytorch/pytorch/blob/master/caffe2/utils/proto_utils.h

hl475 replied:
Aren't those helper functions for c2 operators/arguments? I don't think we have c2 operators/arguments but onnx node/attributes yet upto this line. Please correct me if I am wrong.

houseroad commented on caffe2/onnx/backend.cc:
```diff
@@ -515,6 +515,59 @@ Caffe2Ops Caffe2Backend::CreateConvPoolOpBase(
+  auto pad_a = dummy_->NewDummyName();
+  if (attributes.HasAttribute("count_include_pad")) {
+    auto count_include_pad = attributes.get<int64_t>("count_include_pad", 0L);
+    if (count_include_pad == 1) {
```
You can enhance the check, if all pads are 0, no need to insert PadImage operator.

hl475 commented on caffe2/onnx/backend.cc:
```diff
@@ -515,6 +515,59 @@ Caffe2Ops Caffe2Backend::CreateConvPoolOpBase(
+          "Caffe2 only supports padding 2D Tensor, whereas padding is ", str);
+    }
+    if (count_include_pad == 1 &&
+        not(pads.Get(0) == 0 && pads.Get(1) == 0 && pads.Get(2) == 0 &&
```
!

hl475 commented on caffe2/onnx/backend.cc:
```diff
@@ -515,6 +515,59 @@ Caffe2Ops Caffe2Backend::CreateConvPoolOpBase(
+  // Pad
+  bool padding = false;
+  const std::string pad_name = opset_version < 2 ? "paddings" : "pads";
+  auto pad_a = dummy_->NewDummyName();
```
const auto pad_input

hl475 committed:
Add Support for count_include_pad in AveragePool in Caffe2 ONNX Backend (#9458)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/9458

The goal is to support count_include_pad in Caffe2 ONNX backend. This commit contains the first step - support 4-D tensor cases.
AveragePool with count_include_pad can be expressed as PadImage + AveragePool.

Differential Revision: D8852180

fbshipit-source-id: 1bb9ad484998b6950eadb8c7efbed1e636fb06bf

hl475 commented:
@pytorchbot retest this please

houseroad commented:
@pytorchbot retest this please
